src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -138,8 +138,35 @@ public void On(string methodName, Type[] parameterTypes, Func<object[], Task> ha
 
         private async Task<ReadableChannel<object>> StreamAsyncCore(string methodName, Type returnType, object[] args, CancellationToken cancellationToken)
         {
-            var irq = InvocationRequest.Stream(cancellationToken, returnType, GetNextId(), _loggerFactory, out var channel);
-            await InvokeCore(methodName, irq, args, nonBlocking: false);
+            var invokeCts = new CancellationTokenSource();
+            var irq = InvocationRequest.Stream(invokeCts.Token, returnType, GetNextId(), _loggerFactory, this, out var channel);
+            // After InvokeCore we don't want the irq cancellation token to be triggered.
+            // The stream invocation will be canceled by the CancelInvocationMessage, connection closing, or channel finishing.
+            using (cancellationToken.Register(token => ((CancellationTokenSource)token).Cancel(), invokeCts))
+            {
+                await InvokeCore(methodName, irq, args, nonBlocking: false);
+            }
+
+            if (cancellationToken.CanBeCanceled)
+            {
+                cancellationToken.Register(state =>
+                {
+                    var invocationReq = (InvocationRequest)state;
+                    if (!invocationReq.HubConnection._connectionActive.IsCancellationRequested)
+                    {
+                        // Fire and forget, if it fails that means we aren't connected anymore.
+                        _ = invocationReq.HubConnection.SendHubMessage(new CancelInvocationMessage(invocationReq.InvocationId), invocationReq);
```

Comment:
Log failures here.

Reply:
Failures are logged here https://github.com/aspnet/SignalR/pull/979/files#diff-2e56c8934f766e010c7bd3b60cfdb4e3R235 is that enough?

Reply:
Yep